pkg/virt-launcher/virtwrap/manager.go

```diff
@@ -1179,6 +1146,46 @@ func (l *LibvirtDomainManager) SyncVMI(vmi *v1.VirtualMachineInstance, allowEmul
 	return oldSpec, nil
 }
 
+func (l *LibvirtDomainManager) handleVirDomainState(
```

Comment:
I do not really understand this abstraction.
The actions taken below are perhaps based on the domain state, but attempting to group them together adds no useful structure IMO.

How about grouping the body of each condition into a meaningful operation and converting the if-else into a `switch`?

Reply:
> I do not really understand this abstraction.
The actions taken below are perhaps based on the domain state, but attempting to group them together adds no useful structure IMO.

the diff in SyncVMI shows a good abstraction gain, if you don't agree we can agree to disagree here.

> How about grouping the body of each condition into a meaningful operation and converting the if-else into a switch?

There are a few TODOs there. I'm sure some work is needed to clear it up. We can open a issue for that to tackle it latter (or a good candidate for good-first-issue). My goal with this PR is to reduce the complexity of SyncVMI itself as much as possible to make it easier to introduce libvirtxml.Domain.

Reply:
btw, many thanks @EdDev for reviewing and raising good points here, I'll try to address them in the next hour.

Reply:
> the diff in SyncVMI shows a good abstraction gain, if you don't agree we can agree to disagree here.

My observation is that "handle" tends to group things that are not really strongly related. IMO this is the case here. What does it really mean to "handle virt domain state" in the context of sync?

I find it useful to first split things into small functions as descriptive steps and only then check if there is something that can group them together.

But if you are sure about this, then I will not nag on it further, eventually if another better split is possible, it can still be done later or.

Reply:
When reading the code, my interpretation is that, when SyncVMI() is called, there are actions that need to be taken based on virt domain state. At the moment, they are either to actually start the domain or unpause it.
Moving it to its own function also moves the TODOs that are related to _handling_ virt domain state.

I'm not sure if you feel this shouldn't be factored out as a function that handles virt's domain state or if it is the naming that is bad.

Just to be clear, I'm not strong opined about refactoring code. The goal should be making it easier to read and reduce complexity.. my goal at least. If you think that's not being achieved, it is a no-brainer to me and we can just drop commits or the PR itself.

Reply:
> there are actions that need to be taken based on virt domain state.

That is true, however the hotplug later on are also based on the actual state of the domain.

> At the moment, they are either to actually start the domain or unpause it.

My point is that these two actions together have no informative name, while as separate they do.
My suggestion is to leave these two actions expressed at the caller:
```go
...
switch {
case cli.IsDown(virDomainState) && !vmi.IsRunning() && !vmi.IsFinal():
    if err := startDomain(...); err != nil {...}
case: cli.IsPaused(virDomainState) && !l.paused.contains(vmi.UID):
    if err := resumeDomain(...) err != nil {...}
}
...
```

> we can just drop commits or the PR itself.

I prefer not to drop this PR, please keep it even without this suggestion. Improvements can be done later as well.

Reply:
Okay, instead of moving to a function, I'll move it to a switch instead.